apps/api/src/server.py

```diff
@@ -139,43 +148,79 @@ async def ping():
 
 
 @app.post("/agents")
-def create_agent(agent: AgentBase) -> AgentPublic:
+def create_agent(
+    agent: AgentBase,
+    # Require that the user be signed in, but don't do any other verification
+    user: User = Security(get_user_from_token),
+    # user: User = Depends(get_user_from_token),  # noqa
+) -> AgentPublic:
     """
     Creates an agent. Does not start the agent.
     Only stores it in db.
+    Requires that the user be signed in.
     """
     with Session() as session:
         agent = crud.create_agent(session, agent)
 
         return agent_to_agent_public(agent)
 
 
+def get_agents_conditional_dep(
+    request: Request,
+    user_id: UUID | None = None,
+    user_dynamic_id: UUID | None = None,
+    token: HTTPAuthorizationCredentials | None = Security(optional_jwt),
+) -> User | None:
+    """
+    Makes token optional if neither user_id nor dynamic_id are passed.
+    """
+    if user_id is None and user_dynamic_id is None:
+        return None
+    if not token:
+        raise HTTPException(
+            status_code=401,
+            detail="Token is required if user_id or user_dynamic_id is passed",
+        )
+    return get_user_from_token(
+        request=request,
+        token=token,
+    )
+
+
 @app.get("/agents")
 async def get_agents(
     user_id: UUID | None = None,
     user_dynamic_id: UUID | None = None,
+    user: User | None = Depends(get_agents_conditional_dep),
 ) -> Sequence[AgentPublic]:
     """
     Returns a list of Agents.
     If user_id is passed, returns all agents owned by that user.
     If user_dynamic_id is passed, returns all agents owned by that user.
     If neither is passed, returns all agents.
     Raises a 400 if both user_id and user_dynamic_id are passed.
+    If user query params are passed, valid auth for that user must be provided.
     """
     if user_id and user_dynamic_id:
         raise HTTPException(
             status_code=400,
             detail="Exactly one or zero of user_id or user_dynamic_id may be passed",
         )
-
     with Session() as session:
         if user_dynamic_id:
-            user = crud.get_user_by_dynamic_id(session, user_dynamic_id)
-            if not user:
-                raise HTTPException(status_code=404, detail="User not found")
+            if not user or user_dynamic_id != user.dynamic_id:
```

Comment:
not sure about these. You can see the user's agents by calling this endpoint with no params so it's not really a security issue